File: NotePal/Sources/NotePal/App/PetInteractionManager.swift
import Foundation

@MainActor
final class PetInteractionManager: ObservableObject {
    @Published private(set) var state: PetState = .idle
    @Published private(set) var isHovering = false
    @Published private(set) var bounceToken = 0
    @Published private(set) var celebrateToken = 0

    var onSleep: (() -> Void)?

    private let settings: SettingsStore
    private var sleepTimer: Timer?
    private var temporaryTask: Task<Void, Never>?
    private var panelOpen = false
    private var reminderActive = false

    init(settings: SettingsStore) {
        self.settings = settings
        scheduleSleepTimer()
    }

    func userActivity() {
        if state == .sleeping {
            setTemporaryState(.happy, duration: 1.4)
        }

        scheduleSleepTimer()
    }

    func clicked() {
        userActivity()
        bounceToken += 1
        setTemporaryState(.focused, duration: 0.8)
    }

    func doubleClicked() {
        userActivity()
        bounceToken += 1
        setTemporaryState(.happy, duration: 1.8)
    }

    func rightClicked() {
        userActivity()
        setTemporaryState(.surprised, duration: 1.0)
    }

    func dragging() {
        userActivity()
        state = .focused
    }

    func dragEnded() {
        returnToBaseState()
    }

    func setHovering(_ hovering: Bool) {
        isHovering = hovering
        if hovering {
            userActivity()
        }
    }

    func setPanelOpen(_ isOpen: Bool) {
        panelOpen = isOpen
        userActivity()
        returnToBaseState()
    }

    func reminderStarted() {
        reminderActive = true
        userActivity()
        state = .reminding
    }

    func reminderAcknowledged() {
        reminderActive = false
        userActivity()
        returnToBaseState()
    }

    func todoCompleted() {
        userActivity()
        celebrateToken += 1
        setTemporaryState(.celebrating, duration: 2.0)
    }

    func saveFailed() {
        userActivity()
        setTemporaryState(.surprised, duration: 1.8)
    }

    private func setTemporaryState(_ newState: PetState, duration: TimeInterval) {
        temporaryTask?.cancel()
        state = newState

        temporaryTask = Task { @MainActor [weak self] in
            try? await Task.sleep(for: .milliseconds(Int(duration * 1000)))
            guard !Task.isCancelled else { return }
            self?.returnToBaseState()
        }
    }

    private func returnToBaseState() {
        if reminderActive {
            state = .reminding
        } else if panelOpen {
            state = .thinking
        } else {
            state = .idle
        }
    }

    private func scheduleSleepTimer() {
        sleepTimer?.invalidate()
        sleepTimer = nil

        guard settings.idleToSleepDuration > 0 else {
            return
        }

        let timer = Timer(timeInterval: settings.idleToSleepDuration, repeats: false) { [weak self] _ in
            Task { @MainActor in
                self?.enterSleep()
            }
        }
        timer.tolerance = min(10, settings.idleToSleepDuration * 0.1)
        RunLoop.main.add(timer, forMode: .common)
        sleepTimer = timer
    }

    private func enterSleep() {
        guard !panelOpen, !reminderActive else {
            scheduleSleepTimer()
            return
        }

        temporaryTask?.cancel()
        state = .sleeping
        onSleep?()
    }
}

File: NotePal/Windows/NotePal.Windows/Services/PetInteractionManager.cs
using System.Windows.Threading;

namespace NotePalWindows.Services;

public enum PetState
{
    Idle,
    Happy,
    Thinking,
    Sleeping,
    Surprised,
    Reminding,
    Focused,
    Celebrating
}

public sealed class PetInteractionManager
{
    private readonly SettingsStore _settingsStore;
    private readonly DispatcherTimer _sleepTimer = new();
    private readonly DispatcherTimer _temporaryTimer = new();
    private bool _panelOpen;
    private bool _reminderActive;

    public PetInteractionManager(SettingsStore settingsStore)
    {
        _settingsStore = settingsStore;
        _sleepTimer.Tick += (_, _) => EnterSleep();
        _temporaryTimer.Tick += (_, _) =>
        {
            _temporaryTimer.Stop();
            ReturnToBaseState();
        };
        ScheduleSleepTimer();
    }

    public event EventHandler? Changed;
    public Action? SleepStarted { get; set; }
    public PetState State { get; private set; } = PetState.Idle;
    public bool IsHovering { get; private set; }

    public void UserActivity()
    {
        if (State == PetState.Sleeping)
        {
            SetTemporaryState(PetState.Happy, TimeSpan.FromSeconds(1.4));
        }

        ScheduleSleepTimer();
    }

    public void Clicked()
    {
        UserActivity();
        SetTemporaryState(PetState.Focused, TimeSpan.FromSeconds(0.8));
    }

    public void DoubleClicked()
    {
        UserActivity();
        SetTemporaryState(PetState.Happy, TimeSpan.FromSeconds(1.8));
    }

    public void RightClicked()
    {
        UserActivity();
        SetTemporaryState(PetState.Surprised, TimeSpan.FromSeconds(1.0));
    }

    public void Dragging()
    {
        UserActivity();
        SetState(PetState.Focused);
    }

    public void DragEnded()
    {
        ReturnToBaseState();
    }

    public void SetHovering(bool hovering)
    {
        IsHovering = hovering;
        if (hovering)
        {
            UserActivity();
        }

        Changed?.Invoke(this, EventArgs.Empty);
    }

    public void SetPanelOpen(bool isOpen)
    {
        _panelOpen = isOpen;
        UserActivity();
        ReturnToBaseState();
    }

    public void ReminderStarted()
    {
        _reminderActive = true;
        UserActivity();
        SetState(PetState.Reminding);
    }

    public void ReminderAcknowledged()
    {
        _reminderActive = false;
        UserActivity();
        ReturnToBaseState();
    }

    public void TodoCompleted()
    {
        UserActivity();
        SetTemporaryState(PetState.Celebrating, TimeSpan.FromSeconds(2));
    }

    public void SaveFailed()
    {
        UserActivity();
        SetTemporaryState(PetState.Surprised, TimeSpan.FromSeconds(1.8));
    }

    private void SetTemporaryState(PetState state, TimeSpan duration)
    {
        _temporaryTimer.Stop();
        SetState(state);
        _temporaryTimer.Interval = duration;
        _temporaryTimer.Start();
    }

    private void ReturnToBaseState()
    {
        if (_reminderActive)
        {
            SetState(PetState.Reminding);
        }
        else if (_panelOpen)
        {
            SetState(PetState.Thinking);
        }
        else
        {
            SetState(PetState.Idle);
        }
    }

    private void ScheduleSleepTimer()
    {
        _sleepTimer.Stop();
        if (_settingsStore.IdleToSleepDuration <= TimeSpan.Zero)
        {
            return;
        }

        _sleepTimer.Interval = _settingsStore.IdleToSleepDuration;
        _sleepTimer.Start();
    }

    private void EnterSleep()
    {
        if (_panelOpen || _reminderActive)
        {
            ScheduleSleepTimer();
            return;
        }

        _temporaryTimer.Stop();
        SetState(PetState.Sleeping);
        SleepStarted?.Invoke();
    }

    private void SetState(PetState state)
    {
        State = state;
        Changed?.Invoke(this, EventArgs.Empty);
    }
}
